### DISCUSSION TITLE
PocketBase on Digital Ocean

### DISCUSSION THREAD

**Discussion by hosenur** [pocketbase/pocketbase#1462] [Q&A]
I am hosting PocketBase on DigitalOcean and setting it up manually using nginx and server blocks so that I can run multiple instances in same droplet for my hobby projects, But how do I start the server becase ./pocketbase serve works only until the command is running

**Comment by ganigeorgiev** [comment]
You can run the process in the background (eg. by appending `&`) OR better - register a systemd service(s) that will also take care for auto restarting the process(es) in case of a failure or when you server restarts.

For an example, you can check https://pocketbase.io/docs/going-to-production/#systemd-service.

**Reply by hosenur** [reply]
Okay Thanks! but one question, If I am using systemd services, can I use subdomains pointing to my pocketbase instance?

**Reply by ganigeorgiev** [reply]
You can use a single systemd "template" service file (for examples you can check https://serverfault.com/questions/730239/start-n-processes-with-one-systemd-service-file), or create individual systemd services for each of your apps (this could be easier) and just change the port and `pb_data` dir:

```js
# systemctl enable app1.service
# /lib/systemd/system/app1.service
[Unit]
Description = app1

[Service]
Type           = simple
User           = root
Group          = root
LimitNOFILE    = 4096
Restart        = always
RestartSec     = 5s
StandardOutput = append:/path/to/app1/errors.log
StandardError  = append:/path/to/app1/errors.log
ExecStart      = /path/to/app1/pocketbase serve --http="127.0.0.1:8091" --dir="/path/to/app1/pb_data"

[Install]
WantedBy = multi-user.target
```

In your nginx config, you'll bind the reverse proxy to each of the app local ports, for example:

```nginx
server {
    listen 80;
    server_name example.com;
    client_max_body_size 10M;

    location /app1 {
        # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        proxy_read_timeout 360s;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_pass http://127.0.0.1:8091;
    }

    location /app2 {
        # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        proxy_read_timeout 360s;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_pass http://127.0.0.1:8092;
    }
}
```
(the above is just copy-paste and duplicated from the docs and could be simplified; also it uses path locations, but for subdomains it shouldn't be too much different).

**Reply by hosenur** [reply]
`● pocket.service - pocket
     Loaded: loaded (/lib/systemd/system/pocket.service; enabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Sun 2023-01-01 21:31:20 UTC; 2s ago
    Process: 5478 ExecStart=/root/home/nur/PocketBase/pocket/pocketbase serve --http=127.0.0.1:8090 --dir=/root/home/nur/PocketBase/pocket/pb_data (code=exited, status=209/STDOUT)
   Main PID: 5478 (code=exited, status=209/STDOUT)`

I am getting this, can you please advice what's wrong

**Reply by ganigeorgiev** [reply]
@HOSENUR I'm not sure. Is there any more detailed log when you try to start it, eg. `systemctl start pocketbase`? It is possible that there is a syntax error in the template file or the path to the executable is not correct, or the port is already in use.

You can check your error log file for more information.

To verify that your service file is properly formatted, you can also run something like: `systemd-analyze verify pocket.service`.

You can also run `systemctl status pocket.service` to see if there is more info.

**Reply by mhelbich** [reply]
Same here... Not really sure what I could be doing wrong. Running on Ubuntu 20.04 on a Oracle Cloud ARM instance.

```
systemctl status pocketbase
● pocketbase.service - pocketbase
     Loaded: loaded (/lib/systemd/system/pocketbase.service; enabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Thu 2023-01-05 20:42:58 UTC; 3s ago
    Process: 26684 ExecStart=/home/ubuntu/pocketbase/pocketbase serve --http=172.17.0.1:8090 (code=exited, status=209/STDOUT)
   Main PID: 26684 (code=exited, status=209/STDOUT)
        CPU: 987us
```

**Reply by mhelbich** [reply]
Woops, forgot my systemd config file, in case it helps:
```
[Unit]
Description = pocketbase

[Service]
Type           = simple
User           = root
Group          = root
LimitNOFILE    = 4096
Restart        = always
RestartSec     = 5s
StandardOutput = append:/root/pb/errors.log
StandardError  = append:/root/pb/errors.log
ExecStart      = /usr/bin/sudo /home/ubuntu/pocketbase/pocketbase serve --http="172.17.0.1:8090"

[Install]
WantedBy = multi-user.target
```

**Reply by ganigeorgiev** [reply]
@mhelbich Is there any reason why you are using this IP? If you want to bind to all IPv4 interfaces you can try with `--http="0.0.0.0:80"`, but my recommendation would be to use the public ip of your server.

Additionally, you don't need `sudo` in the ExecStart systemd directive because the used user in your case will be root anyway.

To help with the debugging, you can try also to run `cat /root/pb/errors.log` and see if there are any error log messages.

**Reply by mhelbich** [reply]
> @mhelbich Is there any reason why you are using this IP? If you want to bind to all IPv4 interfaces you can try with `--http="0.0.0.0:80"`, but my recommendation would be to use the public ip of your server.

First of all thanks for the reply, that was fast!

Sadly, that doesn't seem to work:
```
./pocketbase serve --http=0.0.0.0:80
> Server started at: http://0.0.0.0:80
  - REST API: http://0.0.0.0:80/api/
  - Admin UI: http://0.0.0.0:80/_/
2023/01/06 12:06:10 listen tcp 0.0.0.0:80: bind: address already in use
```

> Additionally, you don't need `sudo` in the ExecStart systemd directive because the used user in your case will be root anyway.

I only had that left from a potential reply I read elsewhere to try using that. (Sent the wrong systemd config file, I had already changed it back). But thanks for the reminder!

> To help with the debugging, you can try also to run `cat /root/pb/errors.log` and see if there are any error log messages.

Apparently, it hasn't generated a log file (yet?):
```
sudo cat /root/pb/errors.log
cat: /root/pb/errors.log: No such file or directory
```

Anyway, I was using that IP since I'm trying to get PB to work with the NGINX Proxy Manager. I'm rather new to this, so I might be doing something entirely wrong, but let me try to show what I'm doing:
![ngxpm-pb](https://user-images.githubusercontent.com/54327863/211019479-347a53d4-dcf3-4fa5-8620-a512144e137f.png)
The Nginx Proxy Manager has its own Docker Network - and no matter what I've been trying to do with local addresses, etc... I can't seem to be able to access PB over a proxy with those - so I tried to serve PB over that docker network's gateway, which actually turned out to work for me, because I can now access it through my domain without having to expose any ports:
![pb-proxy](https://user-images.githubusercontent.com/54327863/211020524-b16513dc-07da-4503-bce8-b6bc36f9a473.png)

Now for some reason I had gotten problems trying to auto-start Pocketbase with systemd though, like HOSENUR above... I'm not sure if it was related, but for some reason my NGINX Proxy Manager had also stopped working (although I would assume that was my fault...?) so I had to set it up again.

But this time as well, trying to add it to systemd results in the same error for me:
```
systemctl status pocketbase
● pocketbase.service - pocketbase
     Loaded: loaded (/lib/systemd/system/pocketbase.service; enabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Fri 2023-01-06 13:26:53 UTC; 1s ago
    Process: 2654 ExecStart=/home/ubuntu/pocketbase/pocketbase serve --http=172.18.0.1:8090 (code=exited, status=209/STDOUT)
   Main PID: 2654 (code=exited, status=209/STDOUT)
        CPU: 939us
```

I am confused that it works by manually running the pocketbase exe but not with systemd...
Let me know if there's any other info I could provide for debugging, etc...! :)

**Reply by ganigeorgiev** [reply]
If you are getting `listen tcp 0.0.0.0:80: bind: address already in use`, this means that some other application has already binded to the 80 port, most likely the reverse proxy itself. In this case, you can try with a different port, eg. 8090 (you should also use this port in your reverse proxy configuration).

> I am confused that it works by manually running the pocketbase exe but not with systemd...

Are you starting the systemd service from inside the container or from the host?

> Let me know if there's any other info I could provide for debugging, etc...! :)

Other than checking if there are more detailed logs with `journalctl -u pocketbase`, I have no other idea how to debug it. I'm not familiar with Nginx Proxy Manager and to be honest I'm not sure that I understand what you are trying to do and why do you need it in the first place.

**Reply by ganigeorgiev** [reply]
Additionally, if you are running inside a docker container, you may not need the systemd service at all and you can use the [`restart` policy](https://docs.docker.com/config/containers/start-containers-automatically/#use-a-restart-policy) of your PocketBase container instead.

**Reply by mhelbich** [reply]
> Other than checking if there are more detailed logs with `journalctl -u pocketbase`, I have no other idea how to debug it. I'm not familiar with Nginx Proxy Manager and to be honest I'm not sure that I understand what you are trying to do and why do you need it in the first place.

Wow thank you, the `journalctl` command showed me something that I hadn't spotted before:
![journalctl](https://user-images.githubusercontent.com/54327863/211122142-af8f2e55-5fa2-4fe1-bc14-de13dcda71eb.png)

It appears that there were permission errors for the `/root` directory. So to test it with something else, I just dumped the StdOut/StdErr into another folder (in this case the folder where I had pocketbase running) - and now it seems to be working with no problems!

> If you are getting `listen tcp 0.0.0.0:80: bind: address already in use`, this means that some other application has already binded to the 80 port, most likely the reverse proxy itself. In this case, you can try with a different port, eg. 8090 (you should also use this port in your reverse proxy configuration).

I tried this as well/again but to be honest - I guess I just don't quite understand the NGINX Proxy Manager/Docker combo well enough myself.
![0 0 0 0](https://user-images.githubusercontent.com/54327863/211122734-9ae63eea-ba24-48e5-a33d-f5c2c63717d6.png)

I assume the problem here is that the NPM is running in its own docker container (network) and thus has no access to the 0.0.0.0 ip *outside* of its own network.

And `sudo lsof -i -P -n | grep LISTEN` shows the following port bindings:
![lsof](https://user-images.githubusercontent.com/54327863/211122854-7dc5b725-0819-491b-9600-6a84e6e3265c.png)

So anyway, I personally think it's all working fine now. The NPM ip/port issue is something I've fabricated on my own - if I were to run Pocketbase through a docker container, I could (probably) easily just tell it to use NPM's docker network, where it would not have issues with the internal/external network stuff.

And the systemd error through which I had originally found this issue through Google could be solved on my side, primarily - I reckon - by changing the StdOut/StdErr paths. Thanks for the help!

**Comment by xrd** [comment]
@HOSENUR I am running a few pocketbase instances using Dokku (on Digital Ocean). If you want to simplify your setup considerably, you can basically create a Dockerfile and then git push to dokku and have it all running in a minute (and get the benefit of letsencrypt, etc). Dokku is, IMHO, worth investing some time researching. It is basically just a thin wrapper around Docker anyway.

**Reply by nrthbound** [reply]
Thank you for sharing Dokku. I'd never seen it before, and I *hate* Docker with a passion, but this certainly eases the burden of having to use it. Clutch! Thanks again.

**Reply by GD-coding** [reply]
Thanks for this! Getting started with docker and tls certs is challenging for the first time. Dokku seems to be a great help!

**Reply by xrd** [reply]
I'm happy to help. It's an incredibly easy way to get it up and running. Once you know a few commands dokku is so great.

**Comment by gardc** [comment]
Seems like Docker would solve your problems.

**Reply by GD-coding** [reply]
I tried using docker on digitalocean but I am not familiar enough with it. Mostly with https certs. Probably have to set up an nginx proxy or something and that is a lot of setup. If you have any suggestions on how you have done that, that would be amazing.

**Reply by xrd** [reply]
If you install dokku (https://docs.digitalocean.com/products/marketplace/catalog/dokku/) this is everything you need to install pocketbase

```
dokku apps:create pocketbase
dokku git:from-image mydockerurl.com/pocketbase:0.0.1 # You will need to have a docker image pushed to a registry
dokku proxy:set-port http:80:8090 # configure the pocketbase port mapping
dokku letsencrypt:enable # this will add another port mapping for https (https:443:8090) so pocketbase is behind ssl
```

Dokku will handle setting up your nginx proxy in front of pocketbase, will download and setup letsencrypt certs.

Caprover will do this as well. Pick your pleasure!

**Comment by erickythierry** [comment]
I indicate you use Caprover, it is a project that runs on top of Docker Swarm and has everything necessary for you to create containers of your projects. It works similar to Heroku.
With it you already have a nginx configured and integrated with Letsencrypt's Certbot. So every new app you create, you can activate SSL from its web dashboard without a headache.

Now, the best part: in the admin panel of Caprover there is a list of OneClick Apps, and in it has PocketBase for you to install with a few clicks and already have an instance with SSL and ready to use. This way you can create as many instances of PocketBase as you want.

I use Caprover for months without problems, I host several micro services that I use and it is very easy to manage by it.
